utils/s3/client.cc

```diff
@@ -1092,6 +1093,124 @@ data_sink client::make_upload_jumbo_sink(sstring object_name, std::optional<unsi
+    static constexpr size_t _max_buffers_size = 5_MiB;
+    std::deque<claimed_buffer> _buffers;
+    size_t _buffers_size = 0;
+    bool _is_closing = false;
+    seastar::condition_variable _buffer_cv;
+    future<> _filling_fiber = make_ready_future<>();
+
+    future<> make_filling_fiber() {
+        while (true) {
+            try {
+                s3l.trace("Fiber starts cycle for object '{}'", _object_name);
+                if (_is_closing) {
+                    s3l.trace("Abandoning fiber for object '{}'", _object_name);
+                    co_return;
+                }
+                if (_buffers_size >= _max_buffers_size) {
+                    try {
+                        // just in case we miss the wake-up call use timed `when`
+                        co_await _buffer_cv.when(50ms);
+                    } catch (const seastar::condition_variable_timed_out&) {
+                        // just ignore and carry on
+                        continue;
+                    }
+                }
+                auto req = http::request::make("GET", _client->_host, _object_name);
+                auto range_header = format_range_header(_range);
+                req._headers["Range"] = std::move(range_header);
+                s3l.trace("Fiber for object '{}' will make HTTP request within range {}-{}", _object_name, _range.off, _range.len);
+                co_await _client->make_request(
+                    std::move(req),
+                    [this](group_client&, const http::reply& reply, input_stream<char>&& in_) mutable -> future<> {
+                        if (reply._status != http::reply::status_type::ok && reply._status != http::reply::status_type::partial_content) {
+                            s3l.warn("Fiber for object '{}' failed: {}. Exiting", _object_name, reply._status);
+                            throw httpd::unexpected_status_error(reply._status);
+                        }
+                        auto in = std::move(in_);
+                        co_return co_await in.consume([this, start = s3_clock::now()](temporary_buffer<char> buf) -> future<consumption_result<char>> {
+                            auto recap_start = start;
+                            while (_buffers_size < _max_buffers_size && !_is_closing) {
+                                s3l.trace("Fiber for object '{}' will try to download within range {}-{}", _object_name, _range.off, _range.len);
+                                auto buff_size = buf.size();
+                                _range.off += buff_size;
+                                _range.len -= buff_size;
+                                _buffers_size += buff_size;
+                                s3l.trace("Fiber for object '{}' pushes {} bytes buffer", _object_name, buff_size);
+                                _buffers.emplace_back(std::move(buf), co_await _client->claim_memory(buff_size), s3_clock::now() - start);
+                                _buffer_cv.broadcast();
```

Comment:
Are there more than one waiter on the `_buffer_cv`? Even if yes, why waking up _all_ if there's only one buffer appended to the buffer deque?

Reply:
I think it is a leftover from previous iteration, checking